lib/events/azsessions/azsessions.go

```diff
@@ -0,0 +1,531 @@
+			return nil, err
+		}
+		cfg.Log.WithError(err).Debug("Failed to confirm that the " + sessionContainerName + " container exists, attempting creation.")
+		// someone else might've created the container between GetProperties and
+		// Create, so we ignore AlreadyExists
+		if _, err := cErr2(session.Create(ctx, nil)); err != nil && !trace.IsAlreadyExists(err) {
+			if !trace.IsAccessDenied(err) {
+				return nil, err
+			}
+			cfg.Log.WithError(err).Warn(
+				"Could not create the " + sessionContainerName + " container, please ensure it exists or session recordings will not be stored correctly.")
+		}
+	}
+
+	if _, err := cErr2(inprogress.GetProperties(ctx, nil)); err != nil {
+		if !trace.IsNotFound(err) {
+			return nil, err
+		}
+		cfg.Log.WithError(err).Debug("Failed to confirm that the " + inprogressContainerName + " container exists, attempting creation.")
+		// someone else might've created the container between GetProperties and
+		// Create, so we ignore AlreadyExists
+		if _, err := cErr2(inprogress.Create(ctx, nil)); err != nil && !trace.IsAlreadyExists(err) {
+			if !trace.IsAccessDenied(err) {
+				return nil, err
+			}
+			cfg.Log.WithError(err).Warn(
+				"Could not create the " + inprogressContainerName + " container, please ensure it exists or session recordings will not be stored correctly.")
+		}
+	}
+
+	return &Handler{c: cfg, cred: cred, session: session, inprogress: inprogress}, nil
+}
+
+type Handler struct {
+	c          Config
+	cred       azcore.TokenCredential
+	session    *azblob.ContainerClient
+	inprogress *azblob.ContainerClient
+}
+
+var _ events.MultipartHandler = (*Handler)(nil)
+
+// sessionBlob returns a BlockBlobClient for the blob of the recording of the
+// session. Not expected to ever fail.
+func (h *Handler) sessionBlob(sessionID session.ID) (*azblob.BlockBlobClient, error) {
+	client, err := h.session.NewBlockBlobClient(sessionName(sessionID))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return client, nil
+}
+
+// uploadMarkerBlob returns a BlockBlobClient for the marker blob of the stream
+// upload. Not expected to ever fail.
+func (h *Handler) uploadMarkerBlob(upload events.StreamUpload) (*azblob.BlockBlobClient, error) {
+	client, err := h.inprogress.NewBlockBlobClient(uploadMarkerName(upload))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return client, nil
+}
+
+// partBlob returns a BlockBlobClient for the blob of the part of the specified
+// upload, with the given part number. Not expected to ever fail.
+func (h *Handler) partBlob(upload events.StreamUpload, partNumber int64) (*azblob.BlockBlobClient, error) {
+	client, err := h.inprogress.NewBlockBlobClient(partName(upload, partNumber))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return client, nil
+}
+
+func (h *Handler) Upload(ctx context.Context, sessionID session.ID, reader io.Reader) (string, error) {
+	blob, err := h.sessionBlob(sessionID)
+	if err != nil {
+		return "", err
```

Comment:
Are these errors not being trace.Wrap'ed on purpose? There's a bunch of instances where naked errors are returned above and below.

Reply:
Those errors are returned by functions in the package that are already returning trace errors; do we still need to `trace.Wrap` them every time?